WalletWasabi/Blockchain/Analysis/BlockchainAnalyzer.cs

```diff
@@ -3,60 +3,188 @@
-			// even though the duster does not know who it sent the money.
-			// FTR at the time of writing this comment the transaction processor doesn't even let dust to come this far,
-			// but it's better to be prepared for future changes if it's not too computationally expensive.
-			foreach (var newCoin in tx.WalletOutputs.Where(x => x.Amount > DustThreshold))
+			var inputCount = tx.Transaction.Inputs.Count;
+			var outputCount = tx.Transaction.Outputs.Count;
+
+			var ownInputCount = tx.WalletInputs.Count;
+			var ownOutputCount = tx.WalletOutputs.Count;
+
+			if (ownInputCount == 0)
+			{
+				AnalyzeReceive(tx);
+			}
+			else if (inputCount == ownInputCount)
+			{
+				if (outputCount == ownOutputCount)
+				{
+					AnalyzeSelfSpend(tx);
+				}
+				else
+				{
+					AnalyzeNormalSpend(tx);
+				}
+			}
+			else
+			{
+				AnalyzeCoinjoin(tx);
+			}
+
+			AnalyzeClusters(tx);
+		}
+
+		private void AnalyzeCoinjoin(SmartTransaction tx)
+		{
+			AnalyzeWalletInputs(tx, out HashSet<HdPubKey> distinctWalletInputPubKeys, out int newInputAnonset);
+
+			var indistinguishableWalletOutputs = tx
+				.WalletOutputs.GroupBy(x => x.Amount)
+				.ToDictionary(x => x.Key, y => y.Count());
+
+			foreach (var newCoin in tx.WalletOutputs.ToArray())
 			{
-				// Calculate anonymity sets.
 				// Get the anonymity set of i-th output in the transaction.
 				var anonset = tx.Transaction.GetAnonymitySet(newCoin.Index);
 
-				// If we provided inputs to the transaction.
-				if (tx.WalletInputs.Any())
+				// Don't create many anonset if we've provided a lot of equal outputs.
+				anonset -= indistinguishableWalletOutputs[newCoin.Amount] - 1;
+
+				// We should only consider ourselves once in the anonset.
+				anonset += newInputAnonset - 1;
+
+				HdPubKey hdPubKey = newCoin.HdPubKey;
+				if (hdPubKey.AnonymitySet == HdPubKey.DefaultHighAnonymitySet)
+				{
+					// If the new coin's HD pubkey haven't been used yet
+					// then its anonset haven't been set yet.
+					// In that case the acquired anonset does not have to be intersected with the default anonset,
+					// so this coin gets the aquired anonset.
+					hdPubKey.AnonymitySet = anonset;
+				}
+				else if (distinctWalletInputPubKeys.Contains(hdPubKey))
 				{
-					// Take the input that we provided with the smallest anonset.
-					// And add that to the base anonset from the tx.
-					// Our smallest anonset input is the relevant here, because this way the common input ownership heuristic is considered.
-					// Take minus 1, because we do not want to count own into the anonset, so...
-					// If the anonset of our UTXO would be 1, and the smallest anonset of our inputs would be 1, too, then we don't make...
-					// The new UTXO's anonset 2, but only 1.
-					anonset += tx.WalletInputs.Select(x => x.HdPubKey).Min(x => x.AnonymitySet) - 1;
+					// If it's a reuse of an input's pubkey, then intersection punishment is senseless.
```

Comment:
If there's reuse in input and output side, then output side didn't gain, nor lose anonymity. (See `InputOutputSideAddress`.)

Another way to put it: the intersection punishment is for estimating worse case scenario. If I have a tx where my output reuses an input's pubkey, then the exact anonymity set of the output should be the same as the input's. That's the worse case scenario (and the best.) So lowering it even further makes no sense.

> also isn't there an even worse case: when the new output is deterministically linkable (due to consolidation) to several multiple inputs, and reuses the address of one of them?

Correct, and that's also factored in: If you look at the followup code: `hdPubKey.AnonymitySet = newInputAnonset;`, where `newInputAnonset` comes from the already applied intersection punishment on the inputs in the `AnalyzeWalletInputs` method.

> i think this should just be 1 because if this happens things are kind of catastrophic

That'd be a fine rule of thumb, but the above code already handles that more accurately (see my explanation just before that,) so no need to use a hammer here.

>  and perhaps this should contain some sort of debug log event

I would prefer not introducing the responsibility of anomaly detection to this class:

1. Let's assume I put logging here. Then I should probably put logging to 10-15 other places to be consistent.
2. Then to be even more consistent I would want to start adding new conditions just for the sake of anomaly detection, which would worsen performance and readability.
3. Then someone would want to change something in the future. Now the logging messages would make his job harder, because he would want to not change the behavior/he would not want to remove the logging so he'd end up checking conditions just to make sure the same logging is done.

There's something, called [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) which is not always kept by this project, but I think it makes sense to keep it when we're talking about critical and complex code like this is. So I think anomaly detection should be done outside of the anonymity score calculation.